BSP/Src/BSP_I2C.c

```diff
@@ -0,0 +1,180 @@
+
+#include "BSP_I2C.h"
+#include "BSP.h"
+#include "FreeRTOS.h"
+#include "queue.h"
+
+#define I2C_QUEUE_SIZE   64
+#define I2C_ITEM_LENGTH   sizeof(uint8_t)
+
+static StaticQueue_t __DataStaticQueue;
+static StaticQueue_t __MetaStaticQueue;
+static StaticQueue_t __ReceiveStaticQueue;
+
+
+uint8_t __DataBufferStorage[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+uint8_t __MetaBufferStorage[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH * 2];
+uint8_t __ReceiveBufferStorage[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+uint8_t __ReceiveRawData[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
+
+static QueueHandle_t I2C_DataQueue;
+//addresses, registers, and lengths
+static QueueHandle_t I2C_MetaQueue;
+
+static QueueHandle_t I2C_ReceiveQueue; 
+
+BSP_Status BSP_I2C_Init() {
+    //QUEUES
+    I2C_DataQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, __DataBufferStorage, &__DataStaticQueue);
+    I2C_MetaQueue = xQueueCreateStatic(I2C_QUEUE_SIZE * 2, I2C_ITEM_LENGTH, __MetaBufferStorage, &__MetaStaticQueue);
+    I2C_ReceiveQueue = xQueueCreateStatic(I2C_QUEUE_SIZE, I2C_ITEM_LENGTH, __ReceiveBufferStorage, &__ReceiveStaticQueue);
+
+
+    GPIO_InitTypeDef gpio_struct;
+    I2C_HandleTypeDef i2c_struct;
+
+    if (__HAL_RCC_I2C3_IS_CLK_DISABLED()) {
+        __HAL_RCC_I2C3_CLK_ENABLE();
+    }
+    if (__HAL_RCC_GPIOA_IS_CLK_DISABLED()) {
+        __HAL_RCC_GPIOA_CLK_ENABLE();
+    }
+    if (__HAL_RCC_GPIOC_IS_CLK_DISABLED()) {
+        __HAL_RCC_GPIOC_CLK_ENABLE();
+    }
+
+    // i2c config
+    i2c_struct.Init.ClockSpeed = I2C_CLOCK_SPEED;
+    i2c_struct.Init.DutyCycle = I2C_DUTYCYCLE_2;
+    // 7 bit addr mode
+    i2c_struct.Init.OwnAddress1 = 0x50 << 1;
+    i2c_struct.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
+    i2c_struct.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
+    i2c_struct.Init.GeneralCallMode = (1 << 6); 
+    i2c_struct.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
+
+    // gpio config
+    gpio_struct.Pin = GPIO_PIN_8; 
+    gpio_struct.Mode = MODE_AF;
+    gpio_struct.Pull = GPIO_PULLUP;
+    gpio_struct.Speed = GPIO_SPEED_MEDIUM; 
+    gpio_struct.Alternate = GPIO_AF4_I2C3;
+
+    // PA8 is SCL
+    HAL_GPIO_Init(GPIOA, &gpio_struct);
+
+    // PC9 is SDA
+    gpio_struct.Pin = GPIO_PIN_9;
+    HAL_GPIO_Init(GPIOC, &gpio_struct);
+}
+
+/** 
+* @param    hi2c:            to structure that has config data
+* @param    deviceAdd :      target device address
+* @param    pDataBuff :      data buffer
+* @param    len :           amount of data   
+* @return   HAL Status    
+*/
+void I2CWrite(I2C_HandleTypeDef *hi2c, 
+              uint32_t deviceAdd, 
+              uint8_t* pDataBuff, 
+              uint16_t len) {
+    //copy buffer data over for security incase it changes 
+    uint8_t buffer[I2C_QUEUE_SIZE * I2C_ITEM_LENGTH];
```

Comment:
should these buffers be local to function or static to the file?

Reply:
i think local is right